Hi,

I am busy these days and unable to check if these mixins are working properly. Can you please test them?  Also, I can't launch the game with `MixinDispenserBehavior` enabled.
```
Caused by: java.lang.RuntimeException: Mixin transformation of net.minecraft.block.dispenser.DispenserBehavior$8 failed
Caused by: org.spongepowered.asm.mixin.transformer.throwables.MixinTransformerError: An unexpected critical error was encountered
Caused by: org.spongepowered.asm.mixin.throwables.MixinApplyError: Mixin [allstackable.mixins.json:MixinDispenserBehavior] from phase [DEFAULT] in config [allstackable.mixins.json] from mod [allstackable] FAILED during APPLY
Caused by: org.spongepowered.asm.mixin.injection.throwables.InvalidInjectionException: @Inject annotation on dispenseOne specifies a target class 'net/minecraft/block/dispenser/ItemDispenserBehavior', which is not supported [PREINJECT Applicator Phase -> allstackable.mixins.json:MixinDispenserBehavior -> Prepare Injections ->  -> handler$zgl000$dispenseOne(Lnet/minecraft/util/math/BlockPointer;Lnet/minecraft/item/ItemStack;Lorg/spongepowered/asm/mixin/injection/callback/CallbackInfoReturnable;)V -> Parse]
```

Thanks for the contribution!

--- a/src/main/java/online/connlost/allstackable/mixin/MixinAnvilScreenHandler.java
+++ b/src/main/java/online/connlost/allstackable/mixin/MixinAnvilScreenHandler.java
@@ -1,12 +1,15 @@
 package online.connlost.allstackable.mixin;
 
+import online.connlost.allstackable.util.IItemMaxCount;
 import online.connlost.allstackable.util.ItemsHelper;
+import net.minecraft.entity.player.PlayerEntity;
 import net.minecraft.inventory.Inventory;
 import net.minecraft.item.ItemStack;
 import net.minecraft.screen.AnvilScreenHandler;
 import org.spongepowered.asm.mixin.Mixin;
 import org.spongepowered.asm.mixin.injection.At;
 import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.ModifyVariable;
 import org.spongepowered.asm.mixin.injection.Redirect;
 import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
 
@@ -15,20 +18,23 @@ public class MixinAnvilScreenHandler {
     /**
     * 
     * Decrement the input stacks by one instead of blindly setting them
-    * to an empty stack.
+    * to an empty stack. The decrement is only done on the server to avoid
+    * a desync between the server and client that results in renaming breaking.
     * This prevents entire stacks from being deleted at a time.
     * 
     **/
     @Redirect(method = "onTakeOutput", at = @At(value = "INVOKE", target = "Lnet/minecraft/inventory/Inventory;setStack(ILnet/minecraft/item/ItemStack;)V"))
-    private void decrementOne(Inventory inventory, int slot, ItemStack stack) {
+    private void decrementOne(Inventory inventory, int slot, ItemStack stack, PlayerEntity player, ItemStack takenStack) {
         ItemStack originalStack = inventory.getStack(slot);
         if (ItemsHelper.isModified(originalStack) && originalStack.getCount() > 1) {
             if(stack.isEmpty()) {
                 stack = originalStack;
-                stack.decrement(1);
+                if (!player.world.isClient) {
+                    stack.decrement(1);
+                }
             }
         }
-        
+            
         inventory.setStack(slot, stack);
     }
     
@@ -47,17 +53,18 @@ public class MixinAnvilScreenHandler {
     
     /**
     * 
-    * Only output a single item at a time.
-    * This prevents entire stacks from being repaired/enchanted at a time.
-    * 
+    * Only output a single item at a time when the item was originally capped
+    * at 1.
+    * This prevents entire stacks from being repaired/enchanted at a time with
+    * wrong costs. It assumes that if the item was originally stackable, that
+    * its anvil costs are already balanced.
+    *
     **/
-    @Redirect(method = "updateResult", at = @At(value = "INVOKE", target = "Lnet/minecraft/item/ItemStack;copy()Lnet/minecraft/item/ItemStack;", ordinal = 0))
-    public ItemStack copyOne(ItemStack itemStack) {
-        ItemStack stack = itemStack.copy();
-        if (ItemsHelper.isModified(stack) && stack.getCount() > 1) { 
-            if(!stack.isEmpty()) {
-                stack.setCount(1);
-            }
+    @ModifyVariable(method = "updateResult", at = @At("STORE"), ordinal = 0)
+    private ItemStack copyOne(ItemStack stack) {
+        if (ItemsHelper.isModified(stack) && ((IItemMaxCount) stack.getItem()).getVanillaMaxCount() == 1) { 
+            stack = stack.copy();
+            stack.setCount(1);
         }
         return stack;
     }

--- a/src/main/java/online/connlost/allstackable/mixin/MixinServerPlayNetworkHandler.java
+++ b/src/main/java/online/connlost/allstackable/mixin/MixinServerPlayNetworkHandler.java
@@ -8,27 +8,56 @@ import net.minecraft.server.network.ServerPlayerEntity;
 import org.spongepowered.asm.mixin.Mixin;
 import org.spongepowered.asm.mixin.Shadow;
 import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
 import org.spongepowered.asm.mixin.injection.ModifyVariable;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
 
 import java.util.List;
 
 @Mixin(ServerPlayNetworkHandler.class)
 public class MixinServerPlayNetworkHandler {
-    @Shadow 
+    @Shadow
     public ServerPlayerEntity player;
-    
+
     /**
-    * 
+    *
     * Set the size of the new signed books to the size of the old writable books
-    * instead of only making one new signed book regardless of the stack size
+    * instead of only making one new signed book regardless of the stack size.
     * This prevents unexpected item deletion.
-    * 
+    *
     **/
     @ModifyVariable(method = "addBook", at = @At("STORE"), ordinal = 1)
     public ItemStack fixSignedBookCount(ItemStack itemStack2, TextStream.Message title, List<TextStream.Message> pages, int slotId) {
-        if (ItemsHelper.isModified(itemStack2)) { 
-            itemStack2.setCount(player.getInventory().getStack(slotId).getCount());
+        ItemStack originalStack = player.getInventory().getStack(slotId);
+        if (ItemsHelper.isModified(originalStack)) {
+            itemStack2.setCount(originalStack.getCount());
         }
         return itemStack2;
+    }
+
+    /**
+    *
+    * Split the written book into several stacks if it is over its maximum
+    * stack count. This will occur whenever the writable book stack count is
+    * greater than the written book's maximum stack count.
+    *
+    **/
+    @Inject(method = "addBook", at = @At("RETURN"))
+    public void fixSignedBookOverCount(TextStream.Message title, List<TextStream.Message> pages, int slotId, CallbackInfo ci) {
+        ItemStack itemStack2 = player.getInventory().getStack(slotId);
+        if (ItemsHelper.isModified(itemStack2) && (itemStack2.getCount() > itemStack2.getMaxCount())) {
+            ItemStack splitStack = itemStack2.copy();
+            int count = itemStack2.getCount() % itemStack2.getMaxCount();
+            splitStack.setCount(count);
+            itemStack2.decrement(count);
+            ItemsHelper.insertNewItem(player, splitStack);
+            while(itemStack2.getCount() > itemStack2.getMaxCount()) {
+                splitStack = itemStack2.copy();
+                count = itemStack2.getMaxCount();
+                splitStack.setCount(count);
+                itemStack2.decrement(count);
+                ItemsHelper.insertNewItem(player, splitStack);
+            }
+        }
     }
 }

